src/Games/NexusMods.Games.RedEngine/Cyberpunk2077/SortOrder/RedMod/RedModSortableItemProvider.cs

```diff
@@ -0,0 +1,388 @@
+    /// but the order might not be the most up to date regardless.
+    /// </param>
+    public List<string> GetRedModOrder(IDb? db = null)
+    {
+        var dbToUse = db ?? _connection.Db;
+
+        var redMods = RedModLoadoutGroup.All(dbToUse)
+            .Where(g => g.AsLoadoutItemGroup().AsLoadoutItem().LoadoutId == LoadoutId)
+            .ToList();
+
+        var enabledRedMods = redMods
+            .Where(RedModIsEnabled)
+            .Select(RedModFolder)
+            .ToList();
+
+        // Retrieves the order from the database using the passed db
+        // NOTE: depending on the db passed, the order might not be the latest
+        var sortOrder = RetrieveSortableEntries(dbToUse);
+
+        // Sanitize the order, applying it to the redMods in questions
+        var validatedOrder = SynchronizeSortingToItems(redMods, sortOrder, this);
+
+        return validatedOrder
+            .Where(si => enabledRedMods.Any(m => m == si.RedModFolderName))
+            .Select(si => si.RedModFolderName.ToString())
+            .ToList();
+    }
+
+    private async Task UpdateOrderCache()
+    {
+        await _semaphore.WaitAsync();
+        try
+        {
+            var redModsGroups = _redModsGroups.ToList();
+            var oldOrder = _orderCache.Items.OrderBy(item => item.SortIndex);
+
+            // Update the order
+            var stagingList = SynchronizeSortingToItems(redModsGroups, oldOrder.ToList(), this);
+
+            // Update the database
+            await PersistSortableEntries(stagingList);
+
+            // Update the cache
+            _orderCache.Edit(innerCache =>
+                {
+                    innerCache.Clear();
+                    innerCache.AddOrUpdate(stagingList);
+                }
+            );
+        }
+        finally
+        {
+            _semaphore.Release();
+        }
+    }
+
+    private List<RedModSortableItem> RetrieveSortableEntries(IDb? db = null)
+    {
+        var dbToUse = db ?? _connection.Db;
+
+        return RedModSortableEntry.All(dbToUse)
+            .Where(si => si.IsValid() && si.AsSortableEntry().ParentSortOrderId == _sortOrderId)
+            .OrderBy(si => si.AsSortableEntry().SortIndex)
+            .Select(redModSortableItem =>
+                {
+                    var sortableItem = redModSortableItem.AsSortableEntry();
+                    return new RedModSortableItem(this,
+                        sortableItem.SortIndex,
+                        redModSortableItem.RedModFolderName,
+                        // Temp values, will get updated when we load the RedMods
+                        modName: redModSortableItem.RedModFolderName,
+                        isActive: false
+                    );
+                }
+            )
+            .ToList();
+    }
+
+    private static List<RedModSortableItem> SynchronizeSortingToItems(IList<RedModLoadoutGroup.ReadOnly> redModsGroups, List<RedModSortableItem> orderList, RedModSortableItemProvider provider)
```

Comment:
Just make a note, via XML docs or otherwise that this function updates the existing load order to remove
no longer available items and add new items that were previously not present. It's not immediately
obvious until you dig into the actual method body.

May be worth to rename the parameters to `availableRedMods` and `currentLoadOrder` too, for clarity.

Reply:
Good call, I made some adjustments